## Введение в регрессионный анализ<br>Домашнее задание №2

Задание выполнил(а): **Овчаров Иван, БПТ-222**

### Формат сдачи

В текущий ipynb-файл необходимо вписать решения (код и текст с интерпретацией), сохранить изменения, а затем загрузить файл в папку на Dropbox по ссылке на странице курса. Аккаунт Dropbox для этого не требуется.

Текст с интерпретацией должен быть в отдельных ячейках типа *Markdown*, не в виде комментариев к коду, ячейки с текстом должны быть аккуратно оформлены.

Всего за задание можно получить 20 баллов.

### Задание

В этом задании вам предлагается поработать с данными, используемыми в известном исследовании Д.Асемоглу, С.Джонсона и Д.Робинсона, посвящённом эффекту институтов на экономическое благополучие. 

#### Контекст исследования (на английском)

D. Acemoglu, S. Johnson, and J. A. Robinson in the paper “The Colonial Origins of Comparative Development: An Empirical Investigation” (2001) evaluated the effect of institutions on economic performance. According to their theory, current economic performance of former colonies depends on the type of institutions europeans introduced during the process of colonization. The type of institutions, in its own turn, depends on natural conditions in a colony.

If natural conditions in a colony were bad and caused diseases and higher mortality, europeans tended to set up ‘extractive states’ that was used only to transfer resources to the metropolitan country. In such cases colonizers did not develop high-quality institutions and thus, provided no protection for private property (Kongo). In colonies with good conditions europeans tried to settle more thoroughly and replicated European institutions with higher emphasis on the private property and the system of checks and balancies against government expropriation (Canada, New Zealand).

#### Переменные в CSV-файле

* `shortnam` : сокращённое название страны;
* `logpgp95` : логарифм ВВП на душу населения в 1995 году;
* `avexpr` : индекс защищенности от экспроприации государством частного иностранного финансирования, 
усредненное значение за 1985-1995 годы, принимает значение от 0 до 10, где более высокие значения соответствуют меньшему риску;
* `lat_abst` : мера расстояния до экватора, абсолютное значение широты, переведенное в шкалу от 0 до 1, где 0 соответствует экватору;
* `reg`: континент (Africa, America, Asia);
* `africa`: бинарная переменная, где 1 соответствует странам Африки.
* `asia`: бинарная переменная, где 1 соответствует странам Азии.

Загрузите в Python данные из CSV-файла, и сохраните их в датафрейм `dat`. 

In [1]:
import pandas as pd

dat = pd.read_csv('Овчаров Иван.csv')

Чтобы добавить необходимые для дальнейшей работы столбцы `america` и `reg`, запустите следующую ячейку с кодом (работает, если датафрейм назвали `dat`, как написано выше):

In [2]:
dat["america"] = ((dat["africa"] == 0) & (dat["asia"] == 0)).astype(int)
dat_bin = dat.loc[:, ["africa", "asia", "america"]]
dat["reg"] = dat_bin.apply(lambda x: dat_bin.columns[x.argmax()], 
                                               axis = 1).apply(lambda x: x.capitalize())

In [3]:
dat

Unnamed: 0.1  index  Unnamed: 0 shortnam  africa  asia  lat_abst  \
0              0     35          35      JOR       0     1  0.344444   
1              1     11          11      CHL       0     0  0.333333   
2              2     13          13      CIV       1     0  0.088889   
3              3      9           9      BWA       1     0  0.244444   
4              4     63          63      SYR       0     1  0.388889   
..           ...    ...         ...      ...     ...   ...       ...   
72            72     32          32      IRN       0     1  0.355556   
73            73     25          25      GMB       1     0  0.147556   
74            74     52          52      OMN       0     1  0.233333   
75            75     41          41      MAR       1     0  0.355556   
76            76     26          26      GTM       0     0  0.170000   

      avexpr  logpgp95  america      reg  
0   6.772727  8.171882        0     Asia  
1   7.818182  9.336092        1  America  
2   7.000000  7.444249        0   Africa  
3   7.727273  8.855093        0   Africa  
4   5.818182  8.029433        0     Asia  
..       ...       ...      ...      ...  
72  4.772727  8.605204        0     Asia  
73  8.272727  7.272398        0   Africa  
74  7.136364  9.208339        0     Asia  
75  7.090909  8.042378        0   Africa  
76  5.136364  8.294049        1  America  

[77 rows x 11 columns]

### Задание 1 (2 балла)

Предположим, что нас интересует, какой эффект на экономическое благополучие, выраженное через логарифм ВВП на душу населения, оказывают:

* защищенность от экспроприации государством частного иностранного финансирования;
* расстояние до экватора (представленное в виде меры, предложенной авторами исследования). 

Чтобы это выяснить, необходимо построить модель **множественной линейной регрессии**.

Используя возможности библиотеки `statsmodels`, постройте модель линейной регрессии, которая покажет характер связи между указанными показателями. В качестве результата выполнения этого задания:

* приведите строки кода, необходимые для построения модели, и выдачу, полученную после оценки модели;
* приведите краткое описание модели: какая переменная является зависимой, какие – независимыми.

In [4]:
from statsmodels.formula.api import ols

md1 = ols('logpgp95 ~ avexpr + lat_abst', dat).fit()
md1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               logpgp95   R-squared:                       0.482
Model:                            OLS   Adj. R-squared:                  0.468
Method:                 Least Squares   F-statistic:                     34.42
Date:                Sun, 15 Oct 2023   Prob (F-statistic):           2.71e-11
Time:                        16:45:55   Log-Likelihood:                -89.842
No. Observations:                  77   AIC:                             185.7
Df Residuals:                      74   BIC:                             192.7
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      4.6687      0.445     10.483      0.000       3.781       5.556
avexpr         0.4726      0.071      6.620      0.000       0.330       0.615
lat_abst       1.4901      0.726      2.052      0.044       0.043       2.937
==============================================================================
Omnibus:                        4.139   Durbin-Watson:                   2.712
Prob(Omnibus):                  0.126   Jarque-Bera (JB):                4.071
Skew:                          -0.522   Prob(JB):                        0.131
Kurtosis:                       2.579   Cond. No.                         55.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Объяснение:**

`logpgp95` – зависимая переменная

`avexpr` и `lat_abst` – независимые переменные

### Задание 2 (3 балла)

Проинтерпретируйте полученные результаты: 

* приведите интерпретацию оценки коэффициентов при каждой независимой переменной;
* проверьте, можно ли считать оценку коэффициентов статистически значимыми на уровне значимости 1%;
* запишите, какую долю дисперсии зависимой переменной объясняет построенная модель.

**Интерпретация оценок коэффициентов:**

С увеличением значения индекса `avexpr` (который обозначает индекс защищенности от экспроприации) на 1 пункт (при прочих равных) значение логарфима ВВП (`logpgp95`) увеличивается на 0.4726. С увеличением значения индекса `lat_abst` (который является мерой расстояния до экватора) на 0,1 (при прочих равных) значение логарифма ВВП (`logpgp95`) увеличивается на 0.14901.

**Проверка статистической значимости:**

Для уровня значимости a = 0.01 значение t-статистики при коэффициентах должно: 

$$
{t} \not\subset [{-3},{3}]
$$

t-статистика переменной `avexpr` 6.620 $\not\subset [{-3},{3}]$, соответственно статистически значима 

t-статистика переменной `lat_abst` 2.052 $\subset [{-3},{3}]$, соответственно статистически не значима

**Доля дисперсии переменной:**

Найдем значение ${R}^2$:

In [5]:
print(md1.rsquared)

0.48191963241957947


Построенная модель объясняет ≈0.482 дисперсии зависимой переменной.

### Задание 3 (2 балла)

Запишите уравнение модели (полноценное, со всеми необходимыми индексами и «крышечками»). Вычислите, чему равно, согласно полученной модели, среднее ожидаемое значение логарифма ВВП на душу населения для Бангладеш (код страны BGD).

Характеристики Бангладеш:

|   Unnamed: 0 | shortnam   |   africa |   asia |   lat_abst |   avexpr |   logpgp95 |
|-------------:|:-----------|---------:|-------:|-----------:|---------:|-----------:|
|            4 | BGD        |        0 |      1 |   0.266667 |  5.13636 |     6.8773 |


**Уравнение модели:**

$$
\widehat{\text{logpgp95}}_i={4.6687} +{0.4726} \times \text{avexpr}_i + {1.4901} \times \text{lat_abst}_i.
$$

**Среднее ожидаемое значение для Бангладеша:**

In [10]:
print(dat[dat['shortnam'] == 'BGD'])

    Unnamed: 0.1  index  Unnamed: 0 shortnam  africa  asia  lat_abst  \
46            46      4           4      BGD       0     1  0.266667   

      avexpr  logpgp95  america   reg  
46  5.136364  6.877296        0  Asia  


In [13]:
print(4.6687 + 0.4726 * dat['avexpr'].values[46] + 1.4901 * dat['lat_abst'].values[46])

7.4935054138094195


### Задание 4 (2 балла)

Объясните, можно ли считать переменную «защищенность от экспроприации государством» контрольной переменной в контексте данной модели. 

In [16]:
from scipy import stats

stats.pearsonr(dat['lat_abst'], dat['avexpr'])

PearsonRResult(statistic=0.38671180733616134, pvalue=0.0005126043878669478)

«Контрольная переменная» – это та переменная, которая влияет и на зависимую, и на независимую переменную. Смею предположить, что в контексте данной модели контрольной переменной все же будет `lat_abst`.

### Задание 5 (4 балла)

Добавьте в построенную модель переменную «континент». Приведите строки кода, необходимые для построения модели, выдачу, полученную после оценки модели и ответы на следующие вопросы.

**5.1.(1 балл)** Какой континент был выбран Python в качестве базовой категории?

**5.2. (2 балла)** Проинтерпретируйте коэффициент при `reg[T.Asia]`: объясните, что он содержательно означает и прокомментируйте, является ли оценка коэффициента значимо отличной от 0 на 5%-ном уровне значимости. 

**5.3. (1 балл)** Какая модель, текущая или построенная в задании №1, объясняет большую долю дисперсии зависимой переменной? Как вы думаете, почему так происходит?

In [17]:
md2 = ols('logpgp95 ~ avexpr + lat_abst + reg', dat).fit()
md2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               logpgp95   R-squared:                       0.605
Model:                            OLS   Adj. R-squared:                  0.583
Method:                 Least Squares   F-statistic:                     27.60
Date:                Sun, 15 Oct 2023   Prob (F-statistic):           6.70e-14
Time:                        17:44:56   Log-Likelihood:                -79.378
No. Observations:                  77   AIC:                             168.8
Df Residuals:                      72   BIC:                             180.5
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          4.7911      0.400     11.980      0.000       3.994       5.588
reg[T.America]     0.9734      0.209      4.655      0.000       0.557       1.390
reg[T.Asia]        0.6368      0.209      3.054      0.003       0.221       1.052
avexpr             0.3996      0.066      6.033      0.000       0.268       0.532
lat_abst           0.9691      0.662      1.465      0.147      -0.350       2.288
==============================================================================
Omnibus:                        2.393   Durbin-Watson:                   2.665
Prob(Omnibus):                  0.302   Jarque-Bera (JB):                2.375
Skew:                          -0.400   Prob(JB):                        0.305
Kurtosis:                       2.685   Cond. No.                         56.8
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**5.1** В качестве базовой категории была выбрана Африка.

**5.2** Если значение переменной `Reg[T.Asia]` равно 1, то (при прочих равных) значение логарифма увеличивается на 0.6338. `Reg[T.Asia]` – это dummy-переменная (бинарная, принимающая значения 0 и 1), которая показывает разницу значений логарифма ВВП между странами базовой категории (в нашем случае: странами Африки) и странами Азии.

In [18]:
print(md2.rsquared)

0.6052235994681527


**5.3** Текущая модель объясняет большую долю дисперсии зависимой переменной. Я думаю, что это связано с тем, что в текущей модели больше независимых переменных (через которые мы, собственно и пытаемся объяснить дисперсию). 

### Задание 6 (7 баллов)

Постройте модель множественной линейной регрессии, которая позволяет оценить, каким образом экономическое благополучие зависит от удалённости от экватора с учётом того, что удалённость от экватора оказывает на экономическое благополучие разный эффект в зависимости от того, находится ли страна в Азии или нет. 

**(1 балл)** Приведите строки кода, необходимые для построения модели, выдачу, полученную после оценки модели и ответы на следующие вопросы.

**6.1.(2 балла)** Запишите, чему равен предельный эффект удалённости от экватора для стран Азии.

**6.2.(2 балла)** Проинтерпретируйте коэффициент при эффекте взаимодействия: объясните, что он содержательно означает. 

**6.3.(2 балла)** Можно ли считать, на 5%-ном уровне значимости, что удалённость от экватора оказывает различный эффект на экономическое благополучие в странах Азии и всех остальных странах? Обоснуйте свой ответ.

In [21]:
md3 = ols('logpgp95 ~ lat_abst + asia + lat_abst * asia', dat).fit()
md3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               logpgp95   R-squared:                       0.265
Model:                            OLS   Adj. R-squared:                  0.235
Method:                 Least Squares   F-statistic:                     8.767
Date:                Sun, 15 Oct 2023   Prob (F-statistic):           4.87e-05
Time:                        18:08:45   Log-Likelihood:                -103.31
No. Observations:                  77   AIC:                             214.6
Df Residuals:                      73   BIC:                             224.0
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         7.1050      0.228     31.167      0.000       6.651       7.559
lat_abst          4.2236      1.036      4.077      0.000       2.159       6.288
asia              1.4122      0.482      2.932      0.004       0.452       2.372
lat_abst:asia    -4.0179      1.777     -2.261      0.027      -7.560      -0.476
==============================================================================
Omnibus:                        7.337   Durbin-Watson:                   2.491
Prob(Omnibus):                  0.026   Jarque-Bera (JB):                2.794
Skew:                           0.029   Prob(JB):                        0.247
Kurtosis:                       2.069   Cond. No.                         19.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**6.1** Предельный эффект удаленности от экватора для стран Азии равен -4.0179.

**6.2** Коэффициент при эффекте взаимодействия показывает, как будет изменяться независимая переменная в зависимости от расстояния от экватора при условии, что страна находится в Азии. 

**6.3** Для уровня значимости a = 0.05 значение t-статистики при коэффициентах должно: 

$$
{t} \not\subset [{-2},{2}]
$$

t-статистика переменной `lat_abst:asia` -2.261 $\not\subset [{-2},{2}]$, соответственно статистически значима 
